File: backend/internal/core/service/data_type_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"com.github.hackathon-kifiya.fraud-detection-system/internal/core/domain"
	"com.github.hackathon-kifiya.fraud-detection-system/internal/core/port"
)

var (
	ErrDataTypeNotFound  = errors.New("data type not found")
	ErrReadOnlyMode      = errors.New("data type modifications must be done through the Data Management Service")
	ErrDuplicateDataType = errors.New("data type already exists")
	ErrInvalidSchema     = errors.New("invalid schema definition")
	ErrDataTypeInUse     = errors.New("data type is in use and cannot be deleted")
)

// DataTypeService handles data type operations by delegating to the Data Management Service via repository adapter
type DataTypeService struct {
	repo port.DataTypeRepository
}

// NewDataTypeService creates a new data type service (read-only, fetches from Data Management Service)
func NewDataTypeService(repo port.DataTypeRepository) *DataTypeService {
	return &DataTypeService{
		repo: repo,
	}
}

// CreateDataType is disabled - data types must be created through Data Management Service
func (s *DataTypeService) CreateDataType(ctx context.Context, req domain.CreateDataTypeRequest) (*domain.DataType, error) {
	return nil, ErrReadOnlyMode
}

// GetDataType retrieves a data type by ID
func (s *DataTypeService) GetDataType(ctx context.Context, id string) (*domain.DataType, error) {
	dataType, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, ErrDataTypeNotFound
	}
	return dataType, nil
}

// GetDataTypeByName retrieves a data type by name
func (s *DataTypeService) GetDataTypeByName(ctx context.Context, dataType string) (*domain.DataType, error) {
	dt, err := s.repo.GetByDataType(ctx, dataType)
	if err != nil {
		return nil, ErrDataTypeNotFound
	}
	return dt, nil
}

// ListDataTypes lists all data types with pagination
func (s *DataTypeService) ListDataTypes(ctx context.Context, limit, offset int) (*domain.DataTypeListResponse, error) {
	dataTypes, total, err := s.repo.List(ctx, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("failed to list data types: %w", err)
	}

	return &domain.DataTypeListResponse{
		DataTypes: dataTypes,
		Total:     total,
		Limit:     limit,
		Offset:    offset,
	}, nil
}

// UpdateDataType is disabled - data types must be updated through Data Management Service
func (s *DataTypeService) UpdateDataType(ctx context.Context, id string, req domain.UpdateDataTypeRequest) (*domain.DataType, error) {
	return nil, ErrReadOnlyMode
}

// DeleteDataType is disabled - data types cannot be deleted
func (s *DataTypeService) DeleteDataType(ctx context.Context, id string) error {
	return ErrReadOnlyMode
}

// GetActiveDataTypes retrieves all active data types
func (s *DataTypeService) GetActiveDataTypes(ctx context.Context) ([]domain.DataType, error) {
	dataTypes, err := s.repo.GetActive(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get active data types: %w", err)
	}
	return dataTypes, nil
}

// SearchDataTypes searches data types by query
func (s *DataTypeService) SearchDataTypes(ctx context.Context, query string) ([]domain.DataType, error) {
	dataTypes, err := s.repo.Search(ctx, query)
	if err != nil {
		return nil, fmt.Errorf("failed to search data types: %w", err)
	}
	return dataTypes, nil
}

// The following sync methods are no longer needed - kept for backward compatibility
func (s *DataTypeService) SetRuleEngineSync(client interface{})            {}
func (s *DataTypeService) SetDecisionServiceSync(client interface{})       {}
func (s *DataTypeService) SetDataManagementServiceSync(client interface{}) {}

File: data-management-service/usecases/data_type_service.py
"""Data type service use case."""

from typing import List, Optional
from domain.data_type_models import DataTypeModel, CreateDataTypeRequest, UpdateDataTypeRequest
from domain.repository import DataTypeRepository
import logging

logger = logging.getLogger(__name__)


class DataTypeService:
    """Service for managing data types."""
    
    def __init__(self, repository: DataTypeRepository):
        self.repository = repository
    
    def create_data_type(self, request: CreateDataTypeRequest) -> DataTypeModel:
        """Create a new data type."""
        # Check if data type already exists
        if self.repository.exists(request.data_type):
            raise ValueError(f"Data type '{request.data_type}' already exists")
        
        data_type = DataTypeModel(
            data_type=request.data_type,
            name=request.name,
            description=request.description,
            schema_definition=request.schema_definition,
            sample_data=request.sample_data,
            status=request.status,
            created_by=request.created_by,
            updated_by=request.created_by
        )
        
        return self.repository.create(data_type)
    
    def get_data_type(self, data_type: str) -> Optional[DataTypeModel]:
        """Get a data type by identifier."""
        return self.repository.get_by_id(data_type)
    
    def get_all_data_types(self) -> List[DataTypeModel]:
        """Get all data types."""
        return self.repository.get_all()
    
    def update_data_type(self, data_type: str, request: UpdateDataTypeRequest) -> Optional[DataTypeModel]:
        """Update a data type."""
        existing = self.repository.get_by_id(data_type)
        if not existing:
            raise ValueError(f"Data type '{data_type}' not found")
        
        # Update fields if provided
        if request.name is not None:
            existing.name = request.name
        if request.description is not None:
            existing.description = request.description
        if request.schema_definition is not None:
            existing.schema_definition = request.schema_definition
        if request.sample_data is not None:
            existing.sample_data = request.sample_data
        if request.status is not None:
            existing.status = request.status
        if request.updated_by is not None:
            existing.updated_by = request.updated_by
        
        from datetime import datetime
        existing.updated_at = datetime.now()
        
        return self.repository.update(data_type, existing)
    
    def delete_data_type(self, data_type: str) -> bool:
        """Delete a data type."""
        return self.repository.delete(data_type)

